src/main/java/jcifs/dcerpc/msrpc/lsarpc.java

```diff
@@ -0,0 +1,1046 @@
+
+                _dst = _dst.derive(_namesi);
+                for ( int _i = 0; _i < _namess; _i++ ) {
+                    this.names[ _i ].encode(_dst);
+                }
+            }
+        }
+
+
+        @Override
+        public void decode ( NdrBuffer _src ) throws NdrException {
+            _src.align(4);
+            this.count = _src.dec_ndr_long();
+            int _namesp = _src.dec_ndr_long();
+
+            if ( _namesp != 0 ) {
+                _src = _src.deferred;
+                int _namess = _src.dec_ndr_long();
+                int _namesi = _src.index;
+                _src.advance(16 * _namess);
+
+                if ( this.names == null ) {
+                    if ( _namess < 0 || _namess > 0xFFFF )
+                        throw new NdrException(NdrException.INVALID_CONFORMANCE);
+                    this.names = new LsarTranslatedName[_namess];
+                }
+                _src = _src.derive(_namesi);
+                for ( int _i = 0; _i < _namess; _i++ ) {
+                    if ( this.names[ _i ] == null ) {
+                        this.names[ _i ] = new LsarTranslatedName();
+                    }
+                    this.names[ _i ].decode(_src);
+                }
+            }
+        }
+    }
+
+    public static class LsarClose extends DcerpcMessage {
+
+        @Override
+        public int getOpnum () {
+            return 0x00;
+        }
+
+        public int retval;
+        public rpc.policy_handle handle;
+
+
+        public LsarClose ( rpc.policy_handle handle ) {
+            this.handle = handle;
+        }
+
+
+        @Override
+        public void encode_in ( NdrBuffer _dst ) throws NdrException {
+            this.handle.encode(_dst);
+        }
+
+
+        @Override
+        public void decode_out ( NdrBuffer _src ) throws NdrException {
+            this.handle.decode(_src);
+            this.retval = _src.dec_ndr_long();
+        }
+    }
+
+    public static class LsarQueryInformationPolicy extends DcerpcMessage {
+
+        @Override
+        public int getOpnum () {
+            return 0x07;
+        }
+
+        public int retval;
+        public rpc.policy_handle handle;
+        public short level;
+        public NdrObject info;
+
+
+        public LsarQueryInformationPolicy ( rpc.policy_handle handle, short level, NdrObject info ) {
+            this.handle = handle;
+            this.level = level;
+            this.info = info;
+        }
+
+
+        @Override
+        public void encode_in ( NdrBuffer _dst ) throws NdrException {
+            this.handle.encode(_dst);
+            _dst.enc_ndr_short(this.level);
+        }
+
+
+        @Override
+        public void decode_out ( NdrBuffer _src ) throws NdrException {
+            int _infop = _src.dec_ndr_long();
+            if ( _infop != 0 ) {
+                _src.dec_ndr_short(); /* union discriminant */
+                this.info.decode(_src);
+
+            }
+            this.retval = _src.dec_ndr_long();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 904:910, 1037:1043.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.